lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -964,21 +971,83 @@ class DedupPass : public DedupBase<DedupPass> {
       // Check if there a module with the same hash.
       auto it = moduleHashes.find(h);
       if (it != moduleHashes.end()) {
-        deduper.dedup(it->second, module);
+        auto original = cast<FModuleLike>(it->second);
+        deduper.dedup(original, module);
+        // Record the group ID of the other module.
+        groupMap[module.moduleNameAttr()] = groupMap[original.moduleNameAttr()];
         erasedModules++;
         anythingChanged = true;
         continue;
       }
-
       // Any module not deduplicated must be recorded.
       deduper.record(module);
-
-      // TODO: if the module is marked must dedup, error here.
-
-      // Record the module's hash if it has not been removed.
+      // Add the module to a new dedup group.
+      groupMap[module.moduleNameAttr()] = currentGroup++;
+      // Record the module's hash.
       moduleHashes[h] = module;
     }
 
+    // This part verifies that all modules marked by "MustDedup" have been
+    // properly deduped with each other. For this check to succeed, all modules
+    // have to belong to the same dedup group. If a module is in a different
+    // dedup group, then it was either not deduped or deduped with the wrong
+    // thing.
+
+    // Get the group number from a module path.
+    auto failed = false;
+    auto getGroup = [&](Attribute path) -> unsigned {
+      // Each module is listed as a target "~Circuit|Module" which we have to
+      // parse.
+      auto [lhs, rhs] = path.cast<StringAttr>().getValue().split('|');
+      auto module = StringAttr::get(context, rhs);
+      auto it = groupMap.find(module);
+      if (it == groupMap.end()) {
+        auto diag =
+            circuit->emitError("MustDeduplicateAnnotation references module ")
```

Comment:
We should have all annotations sanity checked before this point so this doesn't happen.  I don't know if we do yet.

Reply:
These ones are especially heinous since I'm parsing a string not a module reference. If they put a "non-local" path it will complain `references module "/a:A" which does not exist` :)